test/known_problems/should_pass/arith_op.erl

```diff
@@ -0,0 +1,8 @@
+-module(arith_op).
+-export([plusintfloat/2, plusfloatint/2]).
+
+-spec plusintfloat(integer(), float()) -> float().
+plusintfloat(X, Y) -> X + Y.
+
+-spec plusfloatint(float(), integer()) -> float().
+plusfloatint(X, Y) -> X + Y.
```

Comment:
We don't have subtyping between `integer()` and `float()` because it is possible to detect whether a number is a float or an integer. E.g.:
`case 5 of 5.0 -> true; 5 -> false end.`

Given that we don't have subtyping I think it's debatable whether these examples should actually pass. It would be somewhat inconsistent to have subtyping only in the case of arithmetic operators and nowhere else.

Reply:
To me it's not debatable that `5.0 + 5` evaluates to `10.0` which has type `float()`. I still haven't read up on the relevant papers (please add to Wiki!) so I generally go by intuition. I get very suspicious when we want to throw away type information (in this case making the codomain wider than it needs to) or make restrictions that make Erlang patterns more similar to a statically typed language (restricting andalso + orelse to work only on booleans).

I really like the spec that @zuiderkwast suggested and it doesn't look to me like we need subtyping to implement it.

Reply:
Yeah, this is not about subtyping as you say. I apologize for the brainfart.

The spec that @zuiderkwast proposed would probably work. I see two risks:

- Performance. When a function with an intersection type is called we go through every type and checks all its arguments with all those types. If we have complex arithmetic expressions we get an exponential blowup here. That makes me nervous.
- Quality of error messages. I haven't checked lately what kind of errors we get from intersection types but it'd be good to make sure that whatever error messages we get are good enough.

The performance problem might be the killer for using the specs. If we implement specialized handling of the types of the arithmetic operators we can make it considerably more efficient.

Reply:
> When a function with an intersection type is called we go through every type and checks all its arguments with all those types. If we have complex arithmetic expressions we get an exponential blowup here. That makes me nervous

Shouldn't we at most reach the type `integer() | float()`? If we can't handle intersection types as simple as these, how can we expect to handle any intersection types? In any case, I'd like us to try this out and see if performance suffers (let's benchmark!).